File: app/src/main/java/i/am/shiro/amai/fragment/StorageSetupFragment.java
package i.am.shiro.amai.fragment;

import android.content.Context;
import android.os.Bundle;
import android.text.format.Formatter;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;

import java.util.List;

import i.am.shiro.amai.Preferences;
import i.am.shiro.amai.R;
import i.am.shiro.amai.model.StorageOption;

import static i.am.shiro.amai.util.LayoutUtil.addChild;

public final class StorageSetupFragment extends Fragment {

    private static final String KEY_SELECTED = "selectedIndex";

    private List<StorageOption> storageOptions;

    private int selectedIndex;

    public StorageSetupFragment() {
        super(R.layout.fragment_storage_setup);
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        storageOptions = StorageOption.getList(context);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (savedInstanceState == null) {
            selectedIndex = 0;
            Preferences.setStoragePath(storageOptions.get(0)
                .getPath());
        } else {
            selectedIndex = savedInstanceState.getInt(KEY_SELECTED, -1);
        }
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt(KEY_SELECTED, selectedIndex);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        view.findViewById(R.id.finishButton)
            .setOnClickListener(v -> onFinish());

        LinearLayout optionsLayout = view.findViewById(R.id.optionsLayout);

        for (int i = 0; i < storageOptions.size(); i++) {
            StorageOption storageOption = storageOptions.get(i);
            int index = i;

            View storageOptionView = addChild(optionsLayout, R.layout.item_storage_option);
            storageOptionView.setSelected(selectedIndex == i);
            storageOptionView.setOnClickListener(v -> {
                optionsLayout.dispatchSetSelected(false);
                storageOptionView.setSelected(true);
                selectedIndex = index;
                Preferences.setStoragePath(storageOption.getPath());
            });

            TextView titleText = storageOptionView.findViewById(R.id.titleText);
            titleText.setText(getString(R.string.storage_label, i));

            TextView pathText = storageOptionView.findViewById(R.id.pathText);
            pathText.setText(storageOption.getPath());

            long spaceFree = storageOption.getSpaceFree();
            String freeSpaceSize = Formatter.formatFileSize(requireContext(), spaceFree);
            TextView freeSpaceText = storageOptionView.findViewById(R.id.freeSpaceText);
            freeSpaceText.setText(getString(R.string.storage_free, freeSpaceSize));

            ProgressBar gauge = storageOptionView.findViewById(R.id.gauge);
            gauge.setProgress(storageOption.getPercentUsed());
        }
    }

    private void onFinish() {
        FragmentManager fragmentManager = requireFragmentManager();
        fragmentManager.popBackStack();
        fragmentManager.beginTransaction()
            .replace(android.R.id.content, new MainFragment())
            .commit();

        Preferences.setFirstRunDone();
    }
}

File: app/src/main/java/i/am/shiro/amai/fragment/StorageSetupFragment.kt
package i.am.shiro.amai.fragment

import android.content.Context
import android.os.Bundle
import android.text.format.Formatter
import android.view.View
import androidx.fragment.app.Fragment
import i.am.shiro.amai.R
import i.am.shiro.amai.dagger.component
import i.am.shiro.amai.databinding.ItemStorageOptionBinding
import i.am.shiro.amai.model.StorageOption
import i.am.shiro.amai.util.addChild
import i.am.shiro.amai.util.goToMain
import i.am.shiro.amai.util.loadInt
import i.am.shiro.amai.util.saveInt
import kotlinx.android.synthetic.main.fragment_storage_setup.*

class StorageSetupFragment : Fragment(R.layout.fragment_storage_setup) {

    private val preferences by lazy { component.preferences }

    private lateinit var storageOptions: List<StorageOption>

    private var selectedIndex: Int = -1

    override fun onAttach(context: Context) {
        super.onAttach(context)

        storageOptions = context.getExternalFilesDirs(null).map { StorageOption(it) }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        savedInstanceState?.loadInt(::selectedIndex)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.saveInt(::selectedIndex)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        finishButton.isEnabled = selectedIndex != -1
        finishButton.setOnClickListener { onFinish() }

        storageOptions.forEachIndexed { i, storageOption ->

            optionsLayout.addChild(ItemStorageOptionBinding::inflate) {
                root.isSelected = selectedIndex == i
                root.setOnClickListener {
                    optionsLayout.dispatchSetSelected(false)
                    root.isSelected = true
                    selectedIndex = i
                    finishButton.isEnabled = true
                }

                titleText.text = getString(R.string.storage_label, i)

                pathText.text = storageOption.path

                val spaceFree = storageOption.spaceFree
                val freeSpaceSize = Formatter.formatFileSize(context, spaceFree)
                freeSpaceText.text = getString(R.string.storage_free, freeSpaceSize)

                gauge.progress = storageOption.percentUsed
            }
        }
    }

    private fun onFinish() {
        goToMain()

        preferences.storagePath = storageOptions[selectedIndex].path
        preferences.isFirstRun = false
    }
}
